[Tooling] Update Fastfile with latest toolkit breaking changes

AliSoftware opened the pull request:
This is part of project ref paaHJt-2sc-p2

⚠️  This PR targets the `release/18.4` branch as I'm hoping to get it merged before this sprint's store submission in order to take the occasion to test it both for 18.4's `finalize_release` and for 18.5's `code_freeze` lanes, so that we can check nothing breaks before officially doing the major release of the toolkit.

---

## What it does

This PR points to the breaking changes introduced in https://github.com/wordpress-mobile/release-toolkit/pull/300, in preparation for the next major version of the release-toolkit.

It updates the `Fastfile` accordingly to fix the call sites of the actions that have been changed in the toolkit:

 - Removes the dummy `app: 'wp_jp'` parameters in calls for those actions.
 - Add a `get_app_name_option!` helper method which checks, for the lanes where the `app:` parameter is still needed (especially the `build_*` "internal" lanes), that the passed parameter is a valid one (either `wordpress` or `jetpack`)
 - Removes the related `# TODO` notes
 - Fix some documentation comments
 - Additional cleanup: move top-level constants like `{WP,JP}_PREFIX` and `{WP,JP}_PACKAGE_NAME` to the dedicated `APP_SPECIFIC_VALUES`

## To Test

As usual with those changes about `Fastfile`, it's super tricky to test without creating side-effects and actual releases.
Some ideas of what we could test still:

 - Run `rubocop` on the `fastlane/Fastfile`, not much to check the style violations (there are many, I actually plan to submit a subsequent PR about this on top of those changes later, but better keep those separate for review-ability), but mostly to check that there are no syntax errors and `rubocop` can read the structure of the file (i.e. there's no missing `end` for each matching `do`, etc)
 - Run `bundle exec fastlane build_internal skip_prechecks:true` and check that the lane complains that the `app:` parameter is missing
 - Run `bundle exec fastlane build_internal app:foo skip_prechecks:true` and check that the lane complains that the `app:` parameter is invalid
 - Run `bundle exec fastlane build_internal app:wordpress skip_prechecks:true` and check that the lane doesn't crash, and lints + builds wordpress Zalpha Debug
 - Run `bundle exec fastlane build_alpha app:wordpress skip_prechecks:true` and check that the lane doesn't crash, and lints + builds wordpress Zalpha Release
 - Run `bundle exec fastlane build_beta app:wordpress skip_prechecks:true` and check that the lane doesn't crash, and lints + builds wordpress Vanilla Release

AliSoftware committed:
Point Pluginfile to WIP release-toolkit branch

AliSoftware committed:
android_betabuild_prechecks: remove app param

AliSoftware committed:
android_build_prechecks: remove app param

AliSoftware committed:
android_bump_version_beta: remove app param

AliSoftware committed:
android_bump_version_final_release: remove app param

AliSoftware committed:
android_bump_version_hotfix: remove app param

AliSoftware committed:
android_bump_version_release: remove app param

AliSoftware committed:
android_codefreeze_prechecks: remove app param

AliSoftware committed:
android_completecodefreeze_prechecks: remove app param

AliSoftware committed:
android_current_branch_is_hotfix: remove app param

AliSoftware committed:
android_finalize_prechecks: remove app param

AliSoftware committed:
android_get_alpha_version: remove app param

AliSoftware committed:
android_get_app_version: remove app param

AliSoftware committed:
android_get_release_version: remove app param

AliSoftware committed:
Fix lane doc comments

AliSoftware committed:
Normalize upload_build_to_play_store lane

AliSoftware committed:
Check and remove TODOs related to `app:` param

AliSoftware committed:
Fix some more lane doc comments

AliSoftware committed:
Introduce get_app_name_option!() to check app param

Checks if the app: parameter is a valid one amongst 'wordpress' or 'jetpack', and throws a user_error! otherwise, to reduce risks of issues with typos

AliSoftware committed:
Fix lane doc comment

AliSoftware committed:
remove spaces around |options| lane param list

AliSoftware committed:
Move app: parameter to front at call sites

For consistency, prefer always passing "app" parameter first, then "version", then "flavor"/"buildType", then the rest.

AliSoftware committed:
Move app-specific global constants into APP_SPECIFIC_VALUES

AliSoftware committed:
fix get_app_name_option! for nil case

AliSoftware committed:
Remove spacing around block params

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -367,11 +359,10 @@ REPOSITORY_NAME="WordPress-Android"
+    UI.user_error!('Please use `finalize_hotfix_release` lane for hotfixes') if android_current_branch_is_hotfix()
 
-    android_finalize_prechecks(options.merge({app: 'wp_jp'}))
+    android_finalize_prechecks(skip_confirm: options[:skip_confirm])
```
ℹ️  For consistency and readability, I took the occasion of passing explicit parameters here (`skip_confirm: options[:skip_confirm]`) rather than passing the whole `options` opaque Hash.

Some might argue that `options` makes it more flexible as if we need to add new options, we won't need to update the call site. Personally, I prefer to be explicit about the parameters passed, both because it makes it look more like a usual function call, but also so that it's clear when your read the `Fastfile` what needs to be passed to those methods and what kind of input data is used.

mokagio replied:
+1, I definitely did the `options.merge` in the past, but I think making the actions more like a function call, as you say, is better in the long run. 

I'd rather go through the "pain" of manually updating every action call site when a new parameter is added and know what's going on than run the risk of other parameters leaking into the action and causing issues the cause of which would be hard to figure out.

AliSoftware commented on fastlane/Pluginfile:
```diff
@@ -6,5 +6,5 @@ group :screenshots, optional: true do
-# Pointing to the result of merging https://github.com/wordpress-mobile/release-toolkit/pull/298 until we're ready to do a breaking release.
-gem 'fastlane-plugin-wpmreleasetoolkit', git: 'git@github.com:wordpress-mobile/release-toolkit.git', ref: '908ff74c45a3ccd022adb028b2b4a2fad296e27a'
+# Pointing to the branch with pending breaking changes just to be sure it works as expected in the clients before doing the major version
+gem 'fastlane-plugin-wpmreleasetoolkit', git: 'git@github.com:wordpress-mobile/release-toolkit.git', branch: 'jetpack/cleanup'
```
ℹ️  I plan to merge this PR as is, test it during `finalize_release` of 18.4 and during upcoming `code_freeze` of 18.5. 

Once we've checked that all works as expected during those and nothing broke, we'll be ready to make an official major version `2.0.0` of the toolkit and point this one to it.

mokagio replied:
I'm happy with this because it has less steps involved. 

Another option that came to mind as I was reading the comment was to ship a `2.0.0-beta.1` version. 

I would feel strongly about using a beta (or alpha) if the toolkit was more of a public facing project. Given we're the only users, we can be more pragmatic 👍

AliSoftware committed:
update release-toolkit to latest commits on branch

mokagio commented on fastlane/Fastfile:
```diff
@@ -424,22 +415,20 @@ REPOSITORY_NAME="WordPress-Android"
   #####################################################################################
   desc "Builds and updates for distribution"
   lane :build_and_upload_release do |options|
-    # TODO: remove 'app' parameter from this toolkit action. For now since it's not used anymore as JP and WP are being unified, use a dummy value to not break API
```
Always good to see a `TODO` comment addressed and removed 😄

mokagio commented on fastlane/Fastfile:
```diff
@@ -1012,6 +997,13 @@ REPOSITORY_NAME="WordPress-Android"
+  def get_app_name_option!(options)
+    app = options[:app]&.downcase
+    UI.user_error!("Missing 'app' parameter. Expected 'app:wordpress' or 'app:jetpack'") if app.nil?
+    UI.user_error!("Invalid 'app' parameter #{app.inspect}. Expected 'wordpress' or 'jetpack'") unless ['wordpress', 'jetpack'].include?(app)
```
TIL about the [`inspect`](https://apidock.com/ruby/String/inspect) method 🎩

AliSoftware replied:
It's actually a method not just on `String` but [on the `Object` parent class](https://apidock.com/ruby/Object/inspect). It's very handy for debugging or printing values, both because it adds quotes around strings but also because it prints `nil` (verbatim) instead of nothing if the variable is `nil`, etc.

mokagio replied:
> It's actually a method not just on String but on the Object parent class.

🔍 `Object` is indeed the first result I got when looking into it. `String` overrides it, though, (which is why I guess it has its own doc page) so that's why I linked it. The "adds quotations" is a handy additional behavior.

mokagio reviewed (approved):
Tested locally with the steps you suggested, sometimes using `app:jetpack` as opposed to `wordpress`, too. Worked as expected<sup>*</sup>

I didn't suggest to remove the `()` from some action calls because of the ongoing RuboCop work, #15442. Easier for the automation to take care of it 😄 

:shipit: 

---

`*`: My `JAVA_HOME` is currently messed up because of a recent Android Studio upgrade I made (_I think that's the reason_), so I didn't really verify that the lanes succeeded. I did get to the point of the `./gradlew` invocation, though, which is enough for me. Any failure from there on would have been in the Gradle realm, which has not been changed in this PR or [its release-toolkit counterpart](https://github.com/wordpress-mobile/release-toolkit/pull/300).

AliSoftware committed:
Update Pluginfile to new reference

Now that https://github.com/wordpress-mobile/release-toolkit/pull/300 has been merged
